File: library/src/com/lisb/android/mediashrink/AudioShrink.java
package com.lisb.android.mediashrink;

import java.nio.ByteBuffer;
import java.util.concurrent.atomic.AtomicReference;

import android.media.MediaCodec;
import android.media.MediaCodecInfo;
import android.media.MediaExtractor;
import android.media.MediaFormat;
import android.media.MediaMuxer;
import android.util.Log;

public class AudioShrink {

	private static final String LOG_TAG = AudioShrink.class.getSimpleName();
	private static final boolean VERBOSE = false;

	private static final long TIMEOUT_USEC = 250;
	private static final String CODEC = "audio/mp4a-latm";
	private static final int AAC_PROFILE = MediaCodecInfo.CodecProfileLevel.AACObjectHE;
	private int bitRate;

	private final MediaExtractor extractor;
	private final MediaMuxer muxer;

	private int sampleCount;

	private OnProgressListener onProgressListener;

	private static final long UPDATE_PROGRESS_INTERVAL_MS = 3 * 1000;

	public AudioShrink(MediaExtractor extractor, MediaMuxer muxer) {
		this.extractor = extractor;
		this.muxer = muxer;
	}

	public void setBitRate(int bitRate) {
		this.bitRate = bitRate;
	}

	public void setOnProgressListener(OnProgressListener onProgressListener) {
		this.onProgressListener = onProgressListener;
	}

	/**
	 * エンコーダの設定に使うフォーマットを作成する。 <br/>
	 * {@link MediaMuxer#addTrack(MediaFormat)} には利用できない。
	 */
	private MediaFormat createEncoderConfigurationFormat(MediaFormat origin) {
		final MediaFormat format = MediaFormat.createAudioFormat(CODEC,
				origin.getInteger(MediaFormat.KEY_SAMPLE_RATE),
				origin.getInteger(MediaFormat.KEY_CHANNEL_COUNT));
		// TODO ビットレートが元の値より大きくならないようにする
		format.setInteger(MediaFormat.KEY_BIT_RATE, bitRate);
		format.setInteger(MediaFormat.KEY_AAC_PROFILE, AAC_PROFILE);

		Log.d(LOG_TAG, "create audio encoder configuration format:" + format);

		return format;
	}

	private void reencode(final int trackIndex, final Integer newTrackIndex,
			final ReencodeListener listener) throws DecodeException {
		final MediaFormat currentFormat = extractor.getTrackFormat(trackIndex);
		final MediaFormat encoderConfigurationFormat = createEncoderConfigurationFormat(currentFormat);

		MediaCodec encoder = null;
		MediaCodec decoder = null;

		ByteBuffer[] decoderInputBuffers = null;
		ByteBuffer[] decoderOutputBuffers = null;
		MediaCodec.BufferInfo decoderOutputBufferInfo = null;

		ByteBuffer[] encoderInputBuffers = null;
		ByteBuffer[] encoderOutputBuffers = null;
		MediaCodec.BufferInfo encoderOutputBufferInfo = null;

		// 進捗取得に利用
		final float durationUs = currentFormat
				.getLong(MediaFormat.KEY_DURATION);
		final long startTimeNs = System.nanoTime();
		long deliverProgressCount = 0;

		try {
			encoder = createEncoder(encoderConfigurationFormat);
			encoder.start();

			encoderInputBuffers = encoder.getInputBuffers();
			encoderOutputBuffers = encoder.getOutputBuffers();
			encoderOutputBufferInfo = new MediaCodec.BufferInfo();

			decoder = createDecoder(currentFormat);

			if (decoder == null) {
				Log.e(LOG_TAG, "audio decoder not found.");
				throw new DecodeException("audio decoder not found.");
			}

			decoder.start();

			decoderInputBuffers = decoder.getInputBuffers();
			decoderOutputBuffers = decoder.getOutputBuffers();
			decoderOutputBufferInfo = new MediaCodec.BufferInfo();

			extractor.selectTrack(trackIndex);

			boolean extractorDone = false;
			boolean decoderDone = false;

			int pendingDecoderOutputBufferIndex = -1;

			long lastExtracterOutputPts = -1;
			long lastDecoderOutputPts = -1;
			long lastEncoderOutputPts = -1;

			while (true) {
				while (!extractorDone) {
					final int decoderInputBufferIndex = decoder
							.dequeueInputBuffer(TIMEOUT_USEC);
					if (decoderInputBufferIndex < 0) {
						break;
					}
					final ByteBuffer decoderInputBuffer = decoderInputBuffers[decoderInputBufferIndex];
					final int size = extractor.readSampleData(
							decoderInputBuffer, 0);

					final long pts = extractor.getSampleTime();
					if (VERBOSE) {
						Log.v(LOG_TAG, "audio extractor output. size:" + size
								+ ", sample time:" + pts + ", sample flags:"
								+ extractor.getSampleFlags());
					}

					if (size >= 0) {
						if (lastExtracterOutputPts >= pts) {
							Log.w(LOG_TAG, "extractor output pts(" + pts
									+ ") is smaller than last pts("
									+ +lastExtracterOutputPts + ").");
						} else {
							lastExtracterOutputPts = pts;
						}

						decoder.queueInputBuffer(decoderInputBufferIndex, 0,
								size, pts, extractor.getSampleFlags());
					}
					extractorDone = !extractor.advance();
					if (extractorDone) {
						Log.d(LOG_TAG, "audio extractor: EOS");
						decoder.queueInputBuffer(decoderInputBufferIndex, 0, 0,
								0, MediaCodec.BUFFER_FLAG_END_OF_STREAM);
					}
					break;
				}

				while (!decoderDone && pendingDecoderOutputBufferIndex == -1) {
					final int decoderOutputBufferIndex = decoder
							.dequeueOutputBuffer(decoderOutputBufferInfo,
									TIMEOUT_USEC);
					if (decoderOutputBufferIndex == MediaCodec.INFO_OUTPUT_BUFFERS_CHANGED) {
						Log.d(LOG_TAG, "audio decoder: output buffers changed");
						decoderOutputBuffers = decoder.getOutputBuffers();
						break;
					}
					if (decoderOutputBufferIndex < 0) {
						break;
					}

					if ((decoderOutputBufferInfo.flags & MediaCodec.BUFFER_FLAG_CODEC_CONFIG) != 0) {
						decoder.releaseOutputBuffer(decoderOutputBufferIndex,
								false);
						break;
					}
					pendingDecoderOutputBufferIndex = decoderOutputBufferIndex;

					final long pts = decoderOutputBufferInfo.presentationTimeUs;
					if (VERBOSE) {
						Log.v(LOG_TAG, "audio decoder output. time:" + pts
								+ ", offset:" + decoderOutputBufferInfo.offset
								+ ", size:" + decoderOutputBufferInfo.size
								+ ", flag:" + decoderOutputBufferInfo.flags);
					}

					if (lastDecoderOutputPts >= pts) {
						Log.w(LOG_TAG, "decoder output pts(" + pts
								+ ") is smaller than last pts("
								+ lastDecoderOutputPts + ").");
					} else {
						lastDecoderOutputPts = pts;
					}

					break;
				}

				while (pendingDecoderOutputBufferIndex != -1) {
					final int encoderInputBufferIndex = encoder
							.dequeueInputBuffer(TIMEOUT_USEC);
					if (encoderInputBufferIndex < 0) {
						break;
					}
					final ByteBuffer encoderInputBuffer = encoderInputBuffers[encoderInputBufferIndex];
					final ByteBuffer decoderOutputBuffer = decoderOutputBuffers[pendingDecoderOutputBufferIndex]
							.duplicate();

					decoderOutputBuffer
							.position(decoderOutputBufferInfo.offset);
					decoderOutputBuffer.limit(decoderOutputBufferInfo.offset
							+ decoderOutputBufferInfo.size);
					encoderInputBuffer.position(0);
					encoderInputBuffer.put(decoderOutputBuffer);

					encoder.queueInputBuffer(encoderInputBufferIndex, 0,
							decoderOutputBufferInfo.size,
							decoderOutputBufferInfo.presentationTimeUs,
							decoderOutputBufferInfo.flags);

					decoder.releaseOutputBuffer(
							pendingDecoderOutputBufferIndex, false);

					if (decoderOutputBufferInfo.size != 0 && listener != null) {
						if (listener.onFrameDecoded(decoder)) {
							return;
						}
					}

					pendingDecoderOutputBufferIndex = -1;
					if ((decoderOutputBufferInfo.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) {
						Log.d(LOG_TAG, "audio decoder: EOS");
						decoderDone = true;
					}
					break;
				}

				while (true) {
					final int encoderOutputBufferIndex = encoder
							.dequeueOutputBuffer(encoderOutputBufferInfo,
									TIMEOUT_USEC);

					if (encoderOutputBufferIndex == MediaCodec.INFO_OUTPUT_BUFFERS_CHANGED) {
						Log.d(LOG_TAG, "audio encoder: output buffers changed");
						encoderOutputBuffers = encoder.getOutputBuffers();
						break;
					}

					if (encoderOutputBufferIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
						// エンコーダに何か入力しないと、ここに来ないエンコーダがあるので注意。
						if (listener != null) {
							if (listener.onEncoderFormatChanged(encoder)) {
								return;
							}
						}
						break;
					}

					if (encoderOutputBufferIndex < 0) {
						break;
					}

					final ByteBuffer encoderOutputBuffer = encoderOutputBuffers[encoderOutputBufferIndex];
					if ((encoderOutputBufferInfo.flags & MediaCodec.BUFFER_FLAG_CODEC_CONFIG) != 0) {
						encoder.releaseOutputBuffer(encoderOutputBufferIndex,
								false);
						break;
					}

					final long pts = encoderOutputBufferInfo.presentationTimeUs;
					if (VERBOSE) {
						Log.v(LOG_TAG, "audio encoder output. time:" + pts
								+ ", offset:" + encoderOutputBufferInfo.offset
								+ ", size:" + encoderOutputBufferInfo.size
								+ ", flag:" + encoderOutputBufferInfo.flags);
					}

					if (encoderOutputBufferInfo.size != 0) {
						if (lastEncoderOutputPts >= pts) {
							Log.w(LOG_TAG, "encoder output pts(" + pts
									+ ") is smaller than last pts("
									+ lastEncoderOutputPts + ").");
						} else {
							if (newTrackIndex != null) {
								muxer.writeSampleData(newTrackIndex,
										encoderOutputBuffer,
										encoderOutputBufferInfo);
								lastEncoderOutputPts = pts;

								// 進捗更新
								if ((System.nanoTime() - startTimeNs) / 1000 / 1000 > UPDATE_PROGRESS_INTERVAL_MS
										* (deliverProgressCount + 1)) {
									deliverProgressCount++;
									if (onProgressListener != null) {
										onProgressListener
												.onProgress((int) (encoderOutputBufferInfo.presentationTimeUs * 100 / durationUs));
									}
								}
							}
						}
					}
					if ((encoderOutputBufferInfo.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) {
						Log.d(LOG_TAG, "audio encoder: EOS");
						return;
					}
					encoder.releaseOutputBuffer(encoderOutputBufferIndex, false);
					break;
				}
			}
		} finally {
			if (encoder != null) {
				encoder.stop();
				encoder.release();
			}

			if (decoder != null) {
				decoder.stop();
				decoder.release();
			}

			extractor.unselectTrack(trackIndex);
		}
	}

	public MediaFormat createOutputFormat(final int trackIndex)
			throws DecodeException {
		sampleCount = 0;
		final AtomicReference<MediaFormat> formatRef = new AtomicReference<>();
		reencode(trackIndex, null, new ReencodeListener() {
			@Override
			public boolean onFrameDecoded(MediaCodec decoder) {
				sampleCount++;
				return false;
			}

			@Override
			public boolean onEncoderFormatChanged(MediaCodec encoder) {
				Log.d(LOG_TAG, "audio encoder: output format changed. "
						+ encoder.getOutputFormat());
				formatRef.set(encoder.getOutputFormat());
				return false;
			}
		});

		if (sampleCount == 0) {
			Log.e(LOG_TAG, "no audio sample found.");
			throw new DecodeException("no audio sample found.");
		}

		return formatRef.get();
	}

	public void shrink(final int trackIndex, final int newTrackIndex)
			throws DecodeException {
		reencode(trackIndex, newTrackIndex, null);
	}

	private MediaCodec createDecoder(final MediaFormat format) {
		final String mimeType = format.getString(MediaFormat.KEY_MIME);
		final MediaCodec decoder = MediaCodec.createByCodecName(Utils
				.selectCodec(mimeType, false).getName());
		if (decoder != null) {
			decoder.configure(format, null, null, 0);

			Log.d(LOG_TAG, "audio decoder:" + decoder.getName());
		}
		return decoder;
	}

	private MediaCodec createEncoder(final MediaFormat format) {
		final MediaCodec encoder = MediaCodec.createByCodecName(Utils
				.selectCodec(CODEC, true).getName());
		encoder.configure(format, null, null, MediaCodec.CONFIGURE_FLAG_ENCODE);
		Log.d(LOG_TAG, "audio encoder:" + encoder.getName());
		return encoder;
	}

	private interface ReencodeListener {
		/**
		 * @return if stop or not
		 */
		boolean onFrameDecoded(MediaCodec decoder);

		/**
		 * @return if stop or not
		 */
		boolean onEncoderFormatChanged(MediaCodec encoder);
	}

}

File: library/src/main/java/com/lisb/android/mediashrink/AudioShrink.kt
package com.lisb.android.mediashrink

import android.annotation.SuppressLint
import android.media.*
import timber.log.Timber
import java.io.IOException
import java.nio.ByteBuffer
import java.util.concurrent.atomic.AtomicReference

class AudioShrink(private val extractor: MediaExtractor,
                  private val muxer: MediaMuxer,
                  private val errorCallback: UnrecoverableErrorCallback) {

    var bitRate = 0
    var onProgressListener: OnProgressListener? = null

    /**
     * エンコーダの設定に使うフォーマットを作成する。 <br></br>
     * [MediaMuxer.addTrack] には利用できない。
     */
    private fun createEncoderConfigurationFormat(decoderOutputFormat: MediaFormat): MediaFormat {
        val channelCount = decoderOutputFormat.getInteger(MediaFormat.KEY_CHANNEL_COUNT)
        val sampleRate = decoderOutputFormat.getInteger(MediaFormat.KEY_SAMPLE_RATE)
        val format = MediaFormat.createAudioFormat(ENCODE_MIMETYPE, sampleRate, channelCount)
        format.setInteger(MediaFormat.KEY_BIT_RATE, bitRate)
        format.setInteger(MediaFormat.KEY_AAC_PROFILE, AAC_PROFILE)
        Timber.tag(TAG).d("create audio encoder configuration format:%s, decoderOutputFormat:%s",
                Utils.toString(format), Utils.toString(decoderOutputFormat))
        return format
    }

    @Throws(DecodeException::class)
    private fun reencode(trackIndex: Int, newTrackIndex: Int?, listener: ReencodeListener?) {
        Timber.tag(TAG).d("reencode. trackIndex:%d, newTrackIndex:%d, isNull(listener):%b",
                trackIndex, newTrackIndex, listener == null)
        // TODO 既存の音声のビットレートが指定された　bitRate よりも低い場合、圧縮せずにそのまま利用する
        val originalFormat = extractor.getTrackFormat(trackIndex)
        var encoder: MediaCodec? = null
        var decoder: MediaCodec? = null
        // 進捗取得に利用
        val durationUs = originalFormat.getLong(MediaFormat.KEY_DURATION).toFloat()
        val startTimeNs = System.nanoTime()
        var deliverProgressCount: Long = 0
        try {
            val encoderOutputBufferInfo = MediaCodec.BufferInfo()
            // create decorder
            decoder = createDecoder(originalFormat)
            if (decoder == null) {
                Timber.tag(TAG).e("audio decoder not found.")
                throw DecodeException("audio decoder not found.")
            }
            decoder.start()
            val decoderOutputBufferInfo = MediaCodec.BufferInfo()
            extractor.selectTrack(trackIndex)
            var extractorDone = false
            var decoderDone = false
            var decoderOutputBuffer: ByteBuffer? = null // エンコーダに入力されていないデータが残っているデコーダの出力
            var decoderOutputBufferIndex: Int = -1
            var lastExtractorOutputPts: Long = -1
            var lastDecoderOutputPts: Long = -1
            var lastEncoderOutputPts: Long = -1
            var sampleCount = 0
            while (true) { // read from extractor, write to decoder
                while (!extractorDone) {
                    val decoderInputBufferIndex = decoder.dequeueInputBuffer(TIMEOUT_USEC)
                    if (decoderInputBufferIndex < 0) {
                        break
                    }
                    val decoderInputBuffer = requireNotNull(decoder.getInputBuffer(decoderInputBufferIndex))
                    val size = extractor.readSampleData(decoderInputBuffer, 0)
                    // extractor.advance() より先に行うこと
                    val pts = extractor.sampleTime
                    var sampleFlags = extractor.sampleFlags
                    Timber.tag(TAG).v("audio extractor output. buffer.pos:%d, buffer.limit:%d, size:%d, sample time:%d, sample flags:%d",
                            decoderInputBuffer.position(), decoderInputBuffer.limit(), size, pts, sampleFlags)
                    extractorDone = !extractor.advance()
                    if (extractorDone) {
                        Timber.tag(TAG).d("audio extractor: EOS, size:%d, sampleCount:%d",
                                size, sampleCount)
                        sampleFlags = sampleFlags or MediaCodec.BUFFER_FLAG_END_OF_STREAM
                        if (sampleCount == 0) {
                            Timber.tag(TAG).e("no audio sample found.")
                            throw DecodeException("no audio sample found.")
                        }
                    }
                    if (size >= 0) {
                        if (lastExtractorOutputPts >= pts) {
                            Timber.tag(TAG).w("extractor output pts(%d) is smaller than last pts(%d)",
                                    pts, lastExtractorOutputPts)
                        } else {
                            lastExtractorOutputPts = pts
                        }
                        decoder.queueInputBuffer(decoderInputBufferIndex, 0, size, pts, sampleFlags)
                    } else if (extractorDone) {
                        decoder.queueInputBuffer(decoderInputBufferIndex, 0, 0, 0, sampleFlags)
                    }
                    break
                }
                // read from decoder
                while (!decoderDone && decoderOutputBuffer == null) {
                    @SuppressLint("WrongConstant")
                    decoderOutputBufferIndex = decoder.dequeueOutputBuffer(
                            decoderOutputBufferInfo, TIMEOUT_USEC)
                    Timber.tag(TAG).v("audio decoder output. bufferIndex:%d, time:%d, offset:%d, size:%d, flags:%d",
                            decoderOutputBufferIndex,
                            decoderOutputBufferInfo.presentationTimeUs,
                            decoderOutputBufferInfo.offset,
                            decoderOutputBufferInfo.size,
                            decoderOutputBufferInfo.flags)
                    if (decoderOutputBufferIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
                        Timber.tag(TAG).d("audio decoder: output format changed. %s",
                                Utils.toString(decoder.outputFormat))
                        break
                    }

                    if (decoderOutputBufferIndex < 0) break

                    // NOTE: flags が MediaCodec.BUFFER_FLAG_CODEC_CONFIG かつ BUFFER_FLAG_END_OF_STREAM
                    // のときがあるので注意。
                    if (decoderOutputBufferInfo.flags and MediaCodec.BUFFER_FLAG_CODEC_CONFIG != 0) {
                        decoder.releaseOutputBuffer(decoderOutputBufferIndex, false)
                        break
                    }
                    decoderOutputBuffer = requireNotNull(decoder.getOutputBuffer(decoderOutputBufferIndex))
                    if (lastDecoderOutputPts >= decoderOutputBufferInfo.presentationTimeUs) {
                        Timber.tag(TAG).w("decoder output pts(%d) is smaller than last pts(%d)",
                                decoderOutputBufferInfo.presentationTimeUs, lastDecoderOutputPts)
                    } else {
                        lastDecoderOutputPts = decoderOutputBufferInfo.presentationTimeUs
                    }
                    break
                }

                // write to encoder
                while (decoderOutputBuffer != null) {
                    if (encoder == null) {
                        // デコーダの制限などでデコード前とデコード後でサンプリングレートやチャネル数が違うことがあるので
                        // Encoder の作成を遅延する。
                        val encoderConfigurationFormat = createEncoderConfigurationFormat(
                                decoder.outputFormat)
                        encoder = createEncoder(encoderConfigurationFormat)
                        encoder.start()
                    }

                    // TODO
                    // デコード後、デコード前と比べてチャネル数やサンプリングレートが上がってしまっていた場合でも
                    // チャネル数やサンプリングレートをデコード前の値に落として圧縮できるようにする

                    val encoderInputBufferIndex = encoder.dequeueInputBuffer(TIMEOUT_USEC)
                    if (encoderInputBufferIndex < 0) {
                        break
                    }
                    val encoderInputBuffer = requireNotNull(encoder.getInputBuffer(encoderInputBufferIndex))
                    encoderInputBuffer.clear()
                    if (encoderInputBuffer.remaining() >= decoderOutputBufferInfo.size) {
                        Timber.tag(TAG).v("audio encoder input. bufferIndex:%d, time:%d, offset:%d, size:%d, flags:%d",
                                encoderInputBufferIndex,
                                decoderOutputBufferInfo.presentationTimeUs,
                                decoderOutputBufferInfo.offset,
                                decoderOutputBufferInfo.size,
                                decoderOutputBufferInfo.flags)
                        decoderOutputBuffer.position(decoderOutputBufferInfo.offset)
                        decoderOutputBuffer.limit(decoderOutputBufferInfo.offset + decoderOutputBufferInfo.size)
                        encoderInputBuffer.put(decoderOutputBuffer)
                        encoder.queueInputBuffer(encoderInputBufferIndex,
                                0,
                                decoderOutputBufferInfo.size,
                                decoderOutputBufferInfo.presentationTimeUs,
                                decoderOutputBufferInfo.flags)
                        decoder.releaseOutputBuffer(decoderOutputBufferIndex, false)
                        if (decoderOutputBufferInfo.size != 0) {
                            sampleCount++
                        }
                        decoderOutputBuffer = null
                        if (decoderOutputBufferInfo.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM != 0) {
                            Timber.tag(TAG).d("audio decoder: EOS")
                            decoderDone = true
                        }
                    } else {
                        // エンコーダの入力バッファがデコーダの出力より小さい
                        val inputBufferRemaining = encoderInputBuffer.remaining()
                        val pts = decoderOutputBufferInfo.presentationTimeUs
                        val flags = decoderOutputBufferInfo.flags.and(MediaCodec.BUFFER_FLAG_END_OF_STREAM.inv())
                        Timber.tag(TAG).v("multi-phase audio encoder input. bufferIndex:%d, time:%d, offset:%d, size:%d, flags:%d",
                                encoderInputBufferIndex,
                                pts,
                                decoderOutputBufferInfo.offset,
                                inputBufferRemaining,
                                flags)
                        decoderOutputBuffer.position(decoderOutputBufferInfo.offset)
                        decoderOutputBuffer.limit(decoderOutputBufferInfo.offset + inputBufferRemaining)
                        encoderInputBuffer.put(decoderOutputBuffer)

                        encoder.queueInputBuffer(encoderInputBufferIndex,
                                0,
                                inputBufferRemaining,
                                pts,
                                flags)
                        sampleCount++

                        decoderOutputBufferInfo.offset += inputBufferRemaining
                        decoderOutputBufferInfo.size -= inputBufferRemaining
                        val outputSamplingRate = decoder.outputFormat.getInteger(MediaFormat.KEY_SAMPLE_RATE)
                        val outputChannelCount = decoder.outputFormat.getInteger(MediaFormat.KEY_CHANNEL_COUNT)
                        val duration = (inputBufferRemaining / 2 / outputChannelCount) * 1000L * 1000L / outputSamplingRate
                        decoderOutputBufferInfo.presentationTimeUs += duration
                    }
                    break
                }
                // write to muxer
                while (encoder != null) {
                    val encoderOutputBufferIndex = encoder.dequeueOutputBuffer(
                            encoderOutputBufferInfo, TIMEOUT_USEC)
                    Timber.tag(TAG).v("audio encoder output. bufferIndex:%d, time:%d, offset:%d, size:%d, flags:%d",
                            encoderOutputBufferIndex,
                            encoderOutputBufferInfo.presentationTimeUs,
                            encoderOutputBufferInfo.offset,
                            encoderOutputBufferInfo.size,
                            encoderOutputBufferInfo.flags)

                    if (encoderOutputBufferIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) { // エンコーダに何か入力しないと、ここに来ないエンコーダがあるので注意。
                        if (listener?.onEncoderFormatChanged(encoder) == true) return
                        break
                    }
                    if (encoderOutputBufferIndex < 0) break

                    val encoderOutputBuffer = requireNotNull(encoder.getOutputBuffer(encoderOutputBufferIndex))
                    // NOTE: flags が MediaCodec.BUFFER_FLAG_CODEC_CONFIG かつ BUFFER_FLAG_END_OF_STREAM
                    // のときがあるので注意。
                    if (encoderOutputBufferInfo.flags and MediaCodec.BUFFER_FLAG_CODEC_CONFIG == 0
                            && encoderOutputBufferInfo.size != 0) {
                        if (lastEncoderOutputPts >= encoderOutputBufferInfo.presentationTimeUs) {
                            Timber.tag(TAG).w("encoder output pts(%d) is smaller than last pts(%d)",
                                    encoderOutputBufferInfo.presentationTimeUs, lastEncoderOutputPts)
                        } else {
                            if (newTrackIndex != null) {
                                muxer.writeSampleData(newTrackIndex, encoderOutputBuffer,
                                        encoderOutputBufferInfo)
                                lastEncoderOutputPts = encoderOutputBufferInfo.presentationTimeUs
                                // 進捗更新
                                if ((System.nanoTime() - startTimeNs) / 1000 / 1000 > UPDATE_PROGRESS_INTERVAL_MS
                                        * (deliverProgressCount + 1)) {
                                    deliverProgressCount++
                                    onProgressListener?.onProgress((encoderOutputBufferInfo.presentationTimeUs * 100 / durationUs).toInt())
                                }
                            }
                        }
                    }
                    encoder.releaseOutputBuffer(encoderOutputBufferIndex, false)
                    if (encoderOutputBufferInfo.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM != 0) {
                        Timber.tag(TAG).d("audio encoder: EOS")
                        return
                    }
                    break
                }
            }
        } catch (e: DecodeException) { // recoverable error
            Timber.tag(TAG).e(e, "Recoverable error occurred on audio shrink.")
            throw e
        } catch (e: Throwable) {
            Timber.tag(TAG).e(e, "Unrecoverable error occurred on audio shrink.")
            errorCallback.onUnrecoverableError(e)
        } finally {
            if (encoder != null) {
                encoder.stop()
                encoder.release()
            }
            if (decoder != null) {
                decoder.stop()
                decoder.release()
            }
            extractor.unselectTrack(trackIndex)
        }
    }

    @Throws(DecodeException::class)
    fun createOutputFormat(trackIndex: Int): MediaFormat {
        val formatRef = AtomicReference<MediaFormat>()
        reencode(trackIndex, null, object : ReencodeListener {
            override fun onEncoderFormatChanged(encoder: MediaCodec): Boolean {
                Timber.tag(TAG).d("audio encoder: output format changed. %s",
                        Utils.toString(encoder.outputFormat))
                formatRef.set(encoder.outputFormat)
                return true
            }
        })
        return formatRef.get()
    }

    @Throws(DecodeException::class)
    fun shrink(trackIndex: Int, newTrackIndex: Int) {
        reencode(trackIndex, newTrackIndex, null)
    }

    @Throws(DecoderCreationException::class)
    private fun createDecoder(format: MediaFormat): MediaCodec? {
        val mimeType = format.getString(MediaFormat.KEY_MIME)
        val codec = Utils.selectCodec(mimeType, false)
        if (codec == null) {
            val detailMessage = "audio decoder codec is not found. mime-type:$mimeType"
            Timber.tag(TAG).e(detailMessage)
            throw DecoderCreationException(detailMessage)
        }
        val codecName = codec.name
        return try {
            // 古いAndroidではNullableだったのでそのまま残している
            @Suppress("RedundantNullableReturnType")
            val decoder: MediaCodec? = MediaCodec.createByCodecName(codecName)
            if (decoder != null) {
                decoder.configure(format, null, null, 0)
                Timber.tag(TAG).d("audio decoder:%s", decoder.name)
            }
            decoder
        } catch (e: IOException) { // later Lollipop.
            val detailMessage = "audio decoder cannot be created. codec-name:$codecName"
            Timber.tag(TAG).e(e, detailMessage)
            throw DecoderCreationException(detailMessage, e)
        } catch (e: IllegalStateException) {
            val detailMessage = "audio decoder cannot be created. codec-name:$codecName"
            Timber.tag(TAG).e(e, detailMessage)
            throw DecoderCreationException(detailMessage, e)
        }
    }

    @Throws(EncoderCreationException::class)
    private fun createEncoder(format: MediaFormat): MediaCodec {
        val codec = Utils.selectCodec(ENCODE_MIMETYPE, true)
        if (codec == null) {
            val detailMessage = "audio encoder codec is not found. media-type:$ENCODE_MIMETYPE"
            Timber.tag(TAG).e(detailMessage)
            throw EncoderCreationException(detailMessage)
        }
        val codecName = codec.name
        return try {
            val encoder = MediaCodec.createByCodecName(codecName)
            encoder.configure(format, null, null, MediaCodec.CONFIGURE_FLAG_ENCODE)
            Timber.tag(TAG).d("audio encoder:%s", encoder.name)
            encoder
        } catch (e: IOException) { // later Lollipop.
            val detailMessage = "audio encoder cannot be created. codec-name:$codecName"
            Timber.tag(TAG).e(e, detailMessage)
            throw EncoderCreationException(detailMessage, e)
        } catch (e: IllegalStateException) { // TODO Change Detail Message If minSDKVersion > 21
            val detailMessage = "audio encoder cannot be created. codec-name:$codecName"
            Timber.tag(TAG).e(e, detailMessage)
            throw EncoderCreationException(detailMessage, e)
        }
    }

    private interface ReencodeListener {
        /**
         * @return if stop or not
         */
        fun onEncoderFormatChanged(encoder: MediaCodec): Boolean
    }

    companion object {
        private const val TAG = "AudioShrink"
        private const val TIMEOUT_USEC: Long = 250
        private const val ENCODE_MIMETYPE = "audio/mp4a-latm"

        // Because AACObjectHE of some encoder(ex. OMX.google.aac.encoder) is buggy,
        // Don't use AACObjectHE.
        //
        // bug example:
        // - specify mono sound but output stereo sound.
        // - output wrong time_base_codec.
        private const val AAC_PROFILE = MediaCodecInfo.CodecProfileLevel.AACObjectLC
        private const val UPDATE_PROGRESS_INTERVAL_MS = 3 * 1000.toLong()
    }

}